pkg/controller/volume/pvcprotection/pvc_protection_controller.go

```diff
@@ -418,3 +437,64 @@ func (c *Controller) enqueuePVCs(pod *v1.Pod, deleted bool) {
 		}
 	}
 }
+
+// This function checks if the PV correspodning to the PVC is still attached to
+// some node.
+func (c *Controller) isVolumeInUseByNode(pvc *v1.PersistentVolumeClaim) (bool, error) {
+	volName := pvc.Spec.VolumeName
+	// isVolumeInUseByNode is called only when PVC is a delete candidate. In the
+	// scenario where a PVC is already marked for deletion and the volume name is
+	// not stamped on the PVC, skip the nodes attachment check.
+	if volName == "" {
+		klog.Infof("Failed to fetch PV name for the PVC %s/%s which is marked for deletion, skip node attached check for PV", pvc.Namespace, pvc.Name)
+		return false, nil
+	}
+
+	// Check node status in the informer cache, if error encountered or volume not found, do not return immediately, try the API server for consistency.
+	if nodeName, ok := pvc.Annotations[AnnSelectedNode]; ok {
```

Comment:
This is not correct. This was the node at provisioning time, but a pod can be rescheduled to different nodes after that.

I think the only way to do this is to scan all node objects. We could potentially create a reverse index to speed up.

Reply:
Ack

Reply:
Another possibility is since we are still processing pods, we can get the nodes from Pod.NodeName.

Longer term though, if we want to replace the Pod check altogether with the Node check, then we would need to do some sort of node listing method. cc @gnufied

Reply:
Do we want to replace the Pod check? I would have thought the common case is the correct pod state is found in the cache.

Although I suppose the expensive part is just querying all the nodes, and so if we're going to do that anyway it would be simpler to have a single check.

Reply:
from the current implementation, it query API server to get the list of pods, I think. And then loop through the each pod spec to match volume name. So I think this is more expensive than checking the nodes (which get it from informer cache) since # of nodes typically smaller # pods.

I am not sure why not use cache to get list of pods in previous implementation, though.

Reply:
I think it does use the cache: https://github.com/kubernetes/kubernetes/blob/c799b926fc3576849089dd85b3deb26867583690/pkg/controller/volume/pvcprotection/pvc_protection_controller.go#L245

Anyway you're right, there will be fewer nodes. And we have to check all pods any time it's not in use which is going to happen at all deletions.

So +1 to replacing the pod check I guess is my short version.

Thanks for all the help

Reply:
In the original design, there were race conditions if we used the cache instead of live lookup. But I believe we solved it by doing a live lookup on kubelet, not the pvc protection controller: https://github.com/kubernetes/community/blame/master/contributors/design-proposals/storage/postpone-pvc-deletion-if-used-in-a-pod.md#L39

If we do eventually want to replace the pod check with a node check, we'll need to go through the exercise again of making sure there's no race conditions due to node + pod objects being out of sync.

Reply:
ha, yes, I was looking at the live lookup here https://github.com/kubernetes/kubernetes/blob/7ceac2baf0820fd354259aa8b7f0e37b0bc6b814/pkg/controller/volume/pvcprotection/pvc_protection_controller.go#L277:22

I am wondering whether it is absolutely necessary. Not matter how hard we try, there might be a time that pod not yet updated when it checks, but after that it is updated.

Reply:
The design relies on the scheduler to stop scheduling pods once the pvc has a deletion timestamp, and for kubelet to reject pods that use pvcs with a deletion timestamp

Reply:
The scheduler only uses cache though, so it can slip through. That is why kubelet does a live lookup of the pvc.

Reply:
Here's more background information on one of the possible races that we addressed in the original design: https://github.com/kubernetes/community/pull/1174#discussion_r144097411